src/App.tsx

```diff
@@ -8,27 +7,104 @@ import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
 
+import { Todo } from './types/Todo';
+import { TodoStatus } from './types/TodoStatus';
+
+import { normalize } from './components/helpers';
+
+import { getTodos } from './api';
+
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [query, setQuery] = useState('');
+  const [status, setStatus] = useState<TodoStatus>(TodoStatus.All);
+  const [isLoaded, setIsLoaded] = useState(false);
+
+  useEffect(() => {
+    getTodos()
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+        setIsLoaded(true);
+      });
+  }, []);
+
+  const selectTodo = (todo: Todo) => {
+    setSelectedTodo(todo);
+  };
+
+  const clearSelectedTodo = () => {
+    setSelectedTodo(null);
+  };
+
+  const filterQuery = (value: string) => {
+    setQuery(value);
+  };
+
+  const selectStatus = (todoStatus: TodoStatus) => {
+    setStatus(todoStatus);
+  };
+
+  const visibleTodos = todos.filter((todo) => {
+    const normalizedQuery = normalize(query);
+    const normalizedTodo = normalize(todo.title);
+
+    const isTodoIncluded = normalizedTodo.includes(normalizedQuery);
+
+    let isStatusMatch: boolean;
+
+    switch (status) {
+      case TodoStatus.Active:
+        isStatusMatch = !todo.completed;
+        break;
+
+      case TodoStatus.Completed:
+        isStatusMatch = todo.completed;
+        break;
+
+      default:
+        isStatusMatch = true;
+    }
+
+    return isTodoIncluded && isStatusMatch;
+  });
```

Comment:
1. Let's move this function to the `helpers` file.
2. You create `const normalizedQuery` every iteration but it doesn't change.
3. Wrap a value of the `visibleTodos` to the `useMemo` hook